engine/gamesys/src/gamesys/test/buffer/metadata.script

```diff
@@ -0,0 +1,148 @@
+    local status, result = pcall(function (thebuffer)
+        local metadata1 = { 1, 2, 3}
+        buffer.set_metadata(thebuffer, hash("metadata-entry"), metadata1, buffer.VALUE_TYPE_INT8)
+        buffer.set_metadata(thebuffer, hash("metadata-entry"), metadata1, buffer.VALUE_TYPE_UINT8)
+    end, self.new_buffer)
+    assert(not status and string.match(result, "RESULT_METADATA_INVALID"), "update metadata with different number type should fail with RESULT_METADATA_INVALID" )
+
+    -- changing the count of number should fail with RESULT_METADATA_INVALID
+    local status, result = pcall(function (thebuffer)
+        local metadata1 = { 1, 2, 3}
+        local metadata2 = { 1, 2, 3, 4}
+        buffer.set_metadata(thebuffer, hash("metadata-entry2"), metadata1, buffer.VALUE_TYPE_INT8)
+        buffer.set_metadata(thebuffer, hash("metadata-entry2"), metadata2, buffer.VALUE_TYPE_INT8)
+    end, self.new_buffer)
+    assert(not status and string.match(result, "RESULT_METADATA_INVALID"), "update metadata with different count of numbers should fail with RESULT_METADATA_INVALID" )
+
+    -- set invalid table with string indices should fail
+    local status, result = pcall(function (thebuffer)
+        local metadata1 = { a=1, b=2}
+        buffer.set_metadata(thebuffer, hash("metadata-entry5"), metadata1, buffer.VALUE_TYPE_INT8)
+        print( buffer.get_metadata(thebuffer, hash("metadata-entry5") ) )
+    end, self.new_buffer)
+    assert(not status)
+
+    -- set empty metadata table should fail
+    local status, result = pcall(function (thebuffer)
+        local metadata1 = {}
+        buffer.set_metadata(thebuffer, hash("metadata-entry5"), metadata1, buffer.VALUE_TYPE_INT8)
+        print( buffer.get_metadata(thebuffer, hash("metadata-entry5") ) )
+    end, self.new_buffer)
+    assert(not status)
+
+    -- retrieving missing metadata entries
+    local metadata, datatype = buffer.get_metadata(self.new_buffer, hash("missing"))
+    assert(metadata == nil, "returned metadata entry should be nil")
+    assert(datatype == nil, "datatype should be nil")
+
+    -- too few arguments for set_metadata
+    local status, result = pcall(function (thebuffer)
+        local metadata1 = { 1, 2, 3}
+        buffer.set_metadata(thebuffer, hash("metadata-entry3"), metadata1)
+    end, self.new_buffer)
+    assert(not status)
+
+    -- too few arguments for get_metadata
+    local status, result = pcall(function (thebuffer)
+        local metadata1 = { 1, 2, 3}
+        buffer.get_metadata(thebuffer)
+    end, self.new_buffer)
+    assert(not status)
+
+    -- bad number type
+    --[[ TODO - this fails with 'LuaStackCheck: Unbalanced Lua stack, expected (4), actual (5)'. why ?
+    local status, result = pcall(function (thebuffer)
+        local metadata1 = { 1, 2, 3}
+        buffer.set_metadata(thebuffer, hash("metadata-entry4"), metadata1,  12345 )
+    end, self.new_buffer)
+    assert(not status)
+    --]]
+
+    test_number_types(self.new_buffer)
+
+end
```

Comment:
When looking after a "get" for "uint64" test, I realized that we'll likely need to add another ticket for adding support for storing this data in a buffer file.

I.e. if they create buffers offline, it makes sense to support storing the meta data at that time too.
But that will have to come later on.

Reply:
it's part of the buffer so I'd say it makes sense.

Shall i create a ticket for this ?